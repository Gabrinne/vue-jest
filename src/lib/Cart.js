import { find } from "lodash";

import { remove } from "lodash";

import Dinero from "dinero.js";

const Money = Dinero;

Money.defaultCurrency = "BRL";
Money.defaultPrecision = 2;

export default class Cart {
  items = [];

  add(item) {
    const itemToFind = { product: item.product };

    if (find(this.items, itemToFind)) {
      remove(this.items, itemToFind);
    }
    return this.items.push(item);
  }

  remove(product) {
    remove(this.items, { product });
  }

  getTotal() {
    return this.items.reduce((acc, item) => {
      return acc.add(Money({ amount: item.quantity * item.product.price }));
    }, Money({ amount: 0 }));
  }

  sumary() {
    const total = this.getTotal().getAmount();
    const items = this.items;

    return {
      total,
      items,
    };
  }

  checkout() {
    const { total, items } = this.sumary();

    this.items = [];

    return {
      total,
      items,
    };
  }
}
