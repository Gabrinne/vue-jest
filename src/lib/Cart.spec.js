import Cart from "./Cart";

describe("Cart", () => {
  let cart;

  let product = {
    title: "Adidas",
    price: 35388,
  };

  let product2 = {
    title: "Nike",
    price: 41872,
  };

  //antes de cada teste
  beforeEach(() => {
    cart = new Cart();
  });

  describe("getTotal()", () => {
    it("Deve retornar 0 quando getTotal() é executado em uma instância recém criada ", () => {
      expect(cart.getTotal().getAmount()).toEqual(0);
    });

    it("deve multiplicar quantidade e preço e retornar a quantia total ", () => {
      cart.add({
        product,
        quantity: 2,
      });

      expect(cart.getTotal().getAmount()).toEqual(70776);
    });

    it("Deve garantir que não exista mais de um produto por vez", () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product,
        quantity: 1,
      });

      expect(cart.getTotal().getAmount()).toEqual(35388);
    });

    it("Deve atualizar o valor total quando o produto é removido ", () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      cart.remove(product);

      expect(cart.getTotal().getAmount()).toEqual(41872);

      //112.648
    });
  });

  describe("checkout()", () => {
    it("deve retornar um objeto com o total e a lista de itens", () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 3,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it("deve retornar um objeto com o total e a lista de itens quando sumary() for chamado", () => {
      cart.add({
        product,
        quantity: 5,
      });

      cart.add({
        product: product2,
        quantity: 3,
      });

      expect(cart.sumary()).toMatchSnapshot();
      expect(cart.getTotal().getAmount()).toBeGreaterThan(0);
    });

    it("deve limpar o carrinho quando checkout() for chamado", () => {
      cart.add({
        product: product2,
        quantity: 3,
      });

      cart.checkout();

      expect(cart.getTotal().getAmount()).toEqual(0);
    });
  });
});
