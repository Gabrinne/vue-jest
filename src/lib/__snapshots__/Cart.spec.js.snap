// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() deve retornar um objeto com o total e a lista de itens 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Nike",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`;

exports[`Cart checkout() deve retornar um objeto com o total e a lista de itens quando sumary() for chamado 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas",
      },
      "quantity": 5,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Nike",
      },
      "quantity": 3,
    },
  ],
  "total": 302556,
}
`;

exports[`Cart checkout() deve retornar um objeto com o total e a lista de itens quando summary() for chamado 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas",
      },
      "quantity": 5,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Nike",
      },
      "quantity": 3,
    },
  ],
  "total": 302556,
}
`;
